### ISSUE TITLE
Clippy wrongly suggest to replace `Vec<Box<T>>` with just `Vec<T>`

### ISSUE THREAD

**Issue by vorner** [rust-lang/rust-clippy#6050]
I tried this code:

```rust
struct Stuff {
    data: Vec<Box<DataThatShouldNotMove>>,
}
```

I have some unsafe code that relies on the instances not changing their address. The `Box` there makes sure that they don't, even if the vec reallocates or when I take them out of the vec and put somewhere else.

Clippy raises the `vec_box` lint and suggests to remove the box.

If I followed clippy's suggestion to replace it with just `Vec<DataThatShouldNotMove>`, the assumption would break.

### Meta

- `cargo clippy -V`: clippy 0.0.212 (dbb73f8 2020-09-12)
- `rustc -Vv`:
  ```
rustc 1.48.0-nightly (dbb73f8f7 2020-09-12)
binary: rustc
commit-hash: dbb73f8f79ab176a897d5a95e696adb71b957cbe
commit-date: 2020-09-12
host: x86_64-unknown-linux-gnu
release: 1.48.0-nightly
LLVM version: 11.0
```

**Comment by flip1995** [comment]
I think you should rather use [`pin`](https://doc.rust-lang.org/std/pin/index.html) for this:

> A `Pin<P>` ensures that the pointee of any pointer type `P` has a stable location in memory, meaning it cannot be moved elsewhere and its memory cannot be deallocated until it gets dropped. We say that the pointee is "pinned".

Anyway, Clippy can't possible detect that the intention behind `Vec<Box<Somthing>>` is, that `Somthing` should not move. For this special case, you should probably only `allow` this lint for that struct (and add a comment, why it is allowed). Or use `Pin`.

**Comment by vorner** [comment]
Well, my point was, there are many possible reasons why one might use a Box and being on the heap is just one of them. But OK :shrug:

**Comment by flip1995** [comment]
Yes, but the elements in the `Vec` are already stored on the heap. That is the point of this lint.
